File: crates/blazen-py/examples/stateful_workflow.py
"""Stateful workflow example.

Demonstrates the two explicit Context namespaces and the
identity-preserving ``StopEvent.result`` feature:

1. ``ctx.state``  -- persistable values that survive ``pause()`` /
   ``resume()``.  Use for counters, JSON-friendly snapshots, anything
   you would want serialised into a checkpoint.
2. ``ctx.session`` -- live in-process references (DB connections, open
   files, lambdas, ML models).  **Identity-preserving** within a run:
   ``ctx.session["conn"]`` returns the *same* Python object across
   steps.  Excluded from snapshots.
3. ``StopEvent(result=obj)`` -- the returned object's ``is``-identity
   is preserved end-to-end, so the caller receives the exact same
   Python instance the workflow produced.

Flow::

    StartEvent -> setup  -> QueryEvent  (writes ctx.state + ctx.session)
    QueryEvent -> query  -> StopEvent   (reads both back, returns conn)

Run with: python stateful_workflow.py
"""

import asyncio
import sqlite3

from blazen import Context, Event, StartEvent, StopEvent, Workflow, step


# ---------------------------------------------------------------------------
# Step 1: Open a sqlite connection, seed it, and hand off.
#
# - ``ctx.state`` holds a persistable counter (would survive a snapshot).
# - ``ctx.session`` holds the live connection (excluded from snapshots,
#   identity preserved across steps).
# ---------------------------------------------------------------------------
@step
async def setup(ctx: Context, ev: StartEvent) -> Event:
    conn = sqlite3.connect(":memory:")
    conn.execute("CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT)")
    conn.executemany(
        "INSERT INTO users (name) VALUES (?)",
        [("alice",), ("bob",), ("carol",)],
    )
    conn.commit()

    # Persistable state -- safe to snapshot.
    ctx.state["row_count_expected"] = 3

    # Live reference -- identity-preserving, not snapshotted.
    ctx.session["db"] = conn

    # Identity is preserved on reads within the same step.
    assert ctx.session["db"] is conn

    print(f"  [setup] stored 3 users; conn id={id(conn)}")
    return Event("QueryEvent")


# ---------------------------------------------------------------------------
# Step 2: Read the live connection back from session and query it.
#
# Demonstrates cross-step identity preservation: the ``conn`` we get
# here is the *same* Python object that ``setup`` stored.
# ---------------------------------------------------------------------------
@step(accepts=["QueryEvent"])
async def query(ctx: Context, ev: Event) -> StopEvent:
    conn = ctx.session["db"]
    (actual_count,) = conn.execute("SELECT count(*) FROM users").fetchone()

    # Update persistable state with the observed count.
    ctx.state["row_count_actual"] = actual_count

    print(f"  [query]  counted {actual_count} users; same conn? {conn is ctx.session['db']}")

    # Return the live connection via StopEvent.result -- identity is
    # preserved end-to-end so the caller receives the exact same object.
    return StopEvent(result=conn)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
async def main() -> None:
    wf = Workflow("stateful-example", [setup, query])

    handler = await wf.run()
    result = await handler.result()

    returned_conn = result.result
    print(f"  [final]  result is sqlite3.Connection? {isinstance(returned_conn, sqlite3.Connection)}")

    # The returned connection is still live -- keep using it.
    (alice_row,) = returned_conn.execute(
        "SELECT name FROM users WHERE id = 1"
    ).fetchone()
    print(f"  [final]  alice row -> {alice_row}")

    returned_conn.close()


if __name__ == "__main__":
    asyncio.run(main())

File: crates/blazen-node/examples/stateful_workflow.ts
/**
 * Stateful Workflow Example
 *
 * Demonstrates the two explicit context namespaces on the Node bindings:
 *   - ctx.state    -- persistable values (survives pause()/resume())
 *   - ctx.session  -- in-process-only values (excluded from snapshots)
 *
 * ---------------------------------------------------------------------------
 * IMPORTANT: Node identity caveat
 * ---------------------------------------------------------------------------
 * On the Node binding, ctx.session values are routed through
 * `serde_json::Value` -- JS object identity is NOT preserved.
 * `await ctx.session.get(key)` returns a plain object equal to the value
 * you passed in, not the same object reference. This is a napi-rs
 * threading limitation: `Reference<T>` is `!Send` because its `Drop`
 * must run on the v8 main thread, so live JS handles cannot be moved
 * into the workflow's Tokio executor.
 *
 * For true identity preservation of live JS objects across steps, use
 * the Python (PyO3) or WASM bindings instead.
 *
 * The session namespace is still functionally distinct from state:
 * session values are excluded from snapshots while state values are not.
 * ---------------------------------------------------------------------------
 *
 * Run with: npx tsx stateful_workflow.ts
 */

import { Workflow } from "blazen";
import type { Context, JsWorkflowResult } from "blazen";

const wf = new Workflow("stateful-example");

// ---------------------------------------------------------------------------
// Step 1: setup -- write to both namespaces.
//
// state: persistable scalar that survives pause()/resume() snapshots.
// session: in-process-only object, NOT snapshotted. Note that the value
// will round-trip through JSON when read back, so identity is lost.
// ---------------------------------------------------------------------------
wf.addStep("setup", ["blazen::StartEvent"], async (event: Record<string, any>, ctx: Context) => {
  // Persistable state -- survives pause/resume.
  await ctx.state.set("row_count_expected", 3);

  // In-process-only values -- excluded from snapshots.
  // Note: identity is NOT preserved; values round-trip through JSON.
  await ctx.session.set("request", { id: "req-abc123", attempt: 1 });

  console.log("[setup] state stored: row_count_expected=3");
  console.log("[setup] session stored: request={id: 'req-abc123', attempt: 1}");
  return { type: "QueryEvent" };
});

// ---------------------------------------------------------------------------
// Step 2: query -- read from both namespaces and prove they are independent.
//
// Demonstrates that state and session are separate keyspaces: writing the
// same key "k" to both namespaces yields two distinct values.
// ---------------------------------------------------------------------------
wf.addStep("query", ["QueryEvent"], async (event: Record<string, any>, ctx: Context) => {
  const expectedCount: number = await ctx.state.get("row_count_expected");
  const request: Record<string, any> = await ctx.session.get("request");
  const hasRequest: boolean = await ctx.session.has("request");

  console.log(`[query] expected_count=${expectedCount}`);
  console.log(`[query] request=${JSON.stringify(request)}, has=${hasRequest}`);

  // Demonstrate state vs session independence: same key in both namespaces.
  await ctx.state.set("k", "state-value");
  await ctx.session.set("k", "session-value");

  return {
    type: "blazen::StopEvent",
    result: {
      expectedCount,
      request,
      hasRequest,
      fromState: await ctx.state.get("k"),
      fromSession: await ctx.session.get("k"),
    },
  };
});

const result: JsWorkflowResult = await wf.run({});

console.log("\nWorkflow finished!");
console.log("  type:", result.type);
console.log("  data:", JSON.stringify(result.data, null, 2));
